File: ch01 . setting/service/api/src/main/java/org/delivery/api/common/api/Result.java
package org.delivery.api.common.api;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.delivery.api.common.error.ErrorCode;
import org.delivery.api.common.error.ErrorCodeInterface;

@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Result {

    private Integer resultCode;

    private String resultMessage;

    private String resultDescription;

    public static Result OK(){
        return Result.builder()
                .resultCode(ErrorCode.OK.getHttpStatusCode())
                .resultMessage(ErrorCode.OK.getDescription())
                .resultDescription("성공")
                .build();
    }

    public static Result ERROR(ErrorCodeInterface errorCodeInterface){
        return Result.builder()
                .resultCode(errorCodeInterface.getErrorCode())
                .resultMessage(errorCodeInterface.getDescription())
                .resultDescription("성공")
                .build();
    }

    public static Result ERROR(ErrorCodeInterface errorCodeInterface, Throwable tx){
        return Result.builder()
                .resultCode(errorCodeInterface.getErrorCode())
                .resultMessage(errorCodeInterface.getDescription())
                .resultDescription(tx.getLocalizedMessage())
                .build();
    }

    public static Result ERROR(ErrorCodeInterface errorCodeInterface, String description){
        return Result.builder()
                .resultCode(errorCodeInterface.getErrorCode())
                .resultMessage(errorCodeInterface.getDescription())
                .resultDescription(description)
                .build();
    }
}

File: ch07 . gateway/service/common/src/main/kotlin/org/delivery/common/api/Result.kt
package org.delivery.common.api

import org.delivery.common.error.ErrorCode
import org.delivery.common.error.ErrorCodeInterface

data class Result(
    val resultCode: Int?=null,
    val resultMessage: String?=null,
    val resultDescription: String?=null,
) {

    companion object { // Java 의 static 과 동일하게 사용
        @JvmStatic
        fun OK(): Result {
            return Result(
                resultCode = ErrorCode.OK.getErrorCode(),
                resultMessage = ErrorCode.OK.getDescription(),
                resultDescription = "성공"
            )
        }
        @JvmStatic
        fun ERROR(errorCodeInterface: ErrorCodeInterface): Result {
            return Result(
                resultCode = errorCodeInterface.getErrorCode(),
                resultMessage = errorCodeInterface.getDescription(),
                resultDescription = "에러 발생"
            )
        }
        @JvmStatic
        fun ERROR(errorCodeInterface: ErrorCodeInterface, tx: Throwable): Result {
            return Result(
                resultCode = errorCodeInterface.getErrorCode(),
                resultMessage = errorCodeInterface.getDescription(),
                resultDescription = tx.localizedMessage
            )
        }
        @JvmStatic
        fun ERROR(errorCodeInterface: ErrorCodeInterface, description: String): Result {
            return Result(
                resultCode = errorCodeInterface.getErrorCode(),
                resultMessage = errorCodeInterface.getDescription(),
                resultDescription = description
            )
        }
    }
}